sdk/src/main/scala/io/horizen/account/state/McAddrOwnershipMsgProcessor.scala

```diff
@@ -525,5 +579,173 @@ object McAddrOwnershipMsgProcessor extends SparkzLogging {
+    val hashedMsg = getMcHashedMsg(Keys.toChecksumAddress(Numeric.toHexString(scAddress.toBytes)))
+
+    // verify MC message signature
+    val recPubKey = Sign.signedMessageHashToKey(hashedMsg, signatureData)
+    val recUncompressedPubKeyBytes = Bytes.concat(Array[Byte](0x04), Numeric.toBytesPadded(recPubKey, PUBLIC_KEY_SIZE))
+    val ecpointRec = ecParameters.getCurve.decodePoint(recUncompressedPubKeyBytes)
+    val recCompressedPubKeyBytes = ecpointRec.getEncoded(true)
+    val mcPubkeyhash = Ripemd160Sha256Hash(recCompressedPubKeyBytes)
+    val computedTaddr = toHorizenPublicKeyAddress(mcPubkeyhash, networkParams)
+
+    computedTaddr.equals(mcTransparentAddress)
+  }
+
+  def isValidOwnershipMultisigSignature(scAddress: Address, mcMultisigAddress: String, pubKey: Array[Byte], mcSignature: SignatureSecp256k1): Boolean = {
+    // get a signature data obj for the verification
+    val v_barr = getUnsignedByteArray(mcSignature.getV)
+    val r_barr = padWithZeroBytes(getUnsignedByteArray(mcSignature.getR), SIGNATURE_RS_SIZE)
+    val s_barr = padWithZeroBytes(getUnsignedByteArray(mcSignature.getS), SIGNATURE_RS_SIZE)
+
+    val signatureData = new Sign.SignatureData(v_barr, r_barr, s_barr)
+
+    // The sc address hex string used in the message to sign must have a checksum format (EIP-55: Mixed-case checksum address encoding)
+    // For a multisig signature we also prepend the string of the mcMultisigAddress in order to prevent the re-utilization of this signature
+    // for adding a standard ownership link with the owner transparent address
+    val hashedMsg = getMcHashedMsg(mcMultisigAddress + Keys.toChecksumAddress(Numeric.toHexString(scAddress.toBytes)))
+
+    // verify MC message signature
+    val recPubKey = Sign.signedMessageHashToKey(hashedMsg, signatureData)
+    val recUncompressedPubKeyBytes = Bytes.concat(Array[Byte](0x04), Numeric.toBytesPadded(recPubKey, PUBLIC_KEY_SIZE))
+
+    val pubKeyBytes = if (pubKey.head == 0x04) {
+      // the format of the input pubKey is uncompressed, no need to get the compressed format
+      recUncompressedPubKeyBytes
+    } else {
+      // get the compressed format
+      val ecpointRec = ecParameters.getCurve.decodePoint(recUncompressedPubKeyBytes)
+      ecpointRec.getEncoded(true)
+    }
+
+    pubKey.sameElements(pubKeyBytes)
+  }
+
+  def verifySignaturesWithThreshold(senderScAddress: Address, mcMultisigAddress: String, pubKeys: Seq[Array[Byte]], signatures: Seq[SignatureSecp256k1], thresholdSignatureValue: Int) : Int = {
+    var score = 0
+    var verifiedPubKeyIndexes = Seq[Int]()
+    val pubKeyIndexPairs = pubKeys.zipWithIndex
+
+    signatures.foreach {
+      signature => {
+        breakable {
+          pubKeyIndexPairs.foreach {
+            pk_idx_pair => {
+              if (score < thresholdSignatureValue && // we are still below the threshold of needed verified signatures
```

Comment:
Maybe we can revert the transaction if one of the signatures is not verified. This will prevent a possible malicious user to add invalid signatures just to waste resources verifying useless signatures.

Reply:
Ok, done